### ISSUE TITLE
[build optimization] Don't build `armeabi-v7a` by default

### ISSUE THREAD

**Issue by qwertychouskie** [FIRST-Tech-Challenge/FtcRobotController#797]
There are no FTC-legal `armeabi-v7a` (aka `arm32`) devices anymore, and EasyOpenCV supports `arm64-v8a` just fine nowadays.  Disabling the build of `armeabi-v7a` cuts the APK from ~60MB to ~50MB.

**Comment by qwertychouskie** [comment]
Also `FtcCommon-9.0.1` includes a bunch of sound files that are not used.  Only these files are actually needed:

- ss_mf_fail.wav
- ss_light_speed.wav
- ss_mine.wav
- ss_bb8_down.wav
- ss_r2d2_up.wav

Everything else is not used, and just eating up space unnecessarily.  Removing these files should shave off another ~3.2MB from the source, and ~2.5MB from the APK download size.

**Comment by qwertychouskie** [comment]
@Windwoes pointed out on Discord that the Moto G5 Plus (which is still competition-legal) runs a 32-bit OS, despite supporting 64-bit hardware.  The other legal phones may also have the same issue.  Depending on how much deploy times benefit, it may still be worth changing the default and having a guide/note to teams on how to re-enable 32-bit support, given that the vast majority of teams use the Control Hub, but this isn't a no-risk change.

**Comment by qwertychouskie** [comment]
Another safe optimization is limiting the formats of fonts bundled to just one or two formats:

<img width="957" height="333" alt="Image" src="https://github.com/user-attachments/assets/9e6edf3a-7f11-42b7-bc19-817006e4754a" />

- SVG fonts are a deprecated standard, and not even supported by most browsers: https://caniuse.com/svg-fonts
- EOT fonts are a non-standard format only supported by Internet Explorer: https://caniuse.com/eot
- Only 1 of the 4 fonts has an OTF
- Only 2 of the 4 fonts has a TTF
- Only 2 of the 4 fonts has a WOFF
- WOFF2 is the only one provided for all 4 fonts.  WOFF2 is also very widely supported, so is probably fine to have as the only format: https://caniuse.com/woff2

I'd recommend just shipping the WOFF2 files, or if Internet Explorer support matters for some reason, only shipping the WOFF files.

**Comment by qwertychouskie** [comment]
I found another ~2.1MB (~40.3MB to ~38.2MB) savings:  The `.map` files for the various web stuff are shipped in the APK despite not being needed by teams (they are basically the web equivalent of debugging symbols, and teams aren't debugging the Javascript and CSS of the web pages hosted by the robot controller).

Excluding them is easy, just add this to the `android` section of `build.common.gradle`:
```gradle
android {

    ...existing stuff...

    // The .map files for the various web stuff are shipped in the APK despite not being needed by
    // teams.  (They are basically the web equivalent of debugging symbols, and teams aren't
    // debugging the Javascript and CSS of the web pages hosted by the robot controller).
    // Remove these files from from the APK to save ~2MB.
    aaptOptions {
        ignoreAssetsPattern "*.map"
    }
}
```

**Comment by qwertychouskie** [comment]
Apparently, you can make Android Studio build separate APKs for armeabi-v7a and arm64-v8a: https://developer.android.com/build/configure-apk-splits#configure-abi-split

This should solve the issue of making the APK smaller, while not breaking the SDK for teams using older RC phones with 32-bit Android versions.

Add this to the `android` section of `build.common.gradle`:
```gradle
android {

    ...existing stuff...

    // Including both arm32 and arm64 binaries in the APK takes up extra space, despite the fact
    // that any given device will only use one of these.  However, at least some of
    // the allowed robot controller phones run 32-bit Android, so we can't just only build arm64.
    // Instead, build separate APKs for arm32 and arm64.  This saves ~8MB from each APK.
    // https://developer.android.com/build/configure-apk-splits#configure-abi-split
    splits {
        abi {
            enable true
            reset()
            include "arm64-v8a", "armeabi-v7a"
            universalApk false
        }
    }
}
```

And remove the two of these from `buildTypes`:
```gradle
ndk {
    abiFilters "armeabi-v7a", "arm64-v8a"
}
```

**Comment by qwertychouskie** [comment]
BTW, I figured out what caused the significant regression in app deployment times in the Into The Deep season, it's an Android Studio regression in 2024.1.2.  Partial deploys worked in 2024.1.1.13, but do not in 2024.1.2.13.  I have a bug report open here: https://issuetracker.google.com/issues/442347342

Hopefully this gets fixed upstream in a timely manner, but any APK size reduction we can do helps mitigate the issue to a certain degree (and even when the upstream partial deploy issue is fixed, a smaller APK is still beneficial, though less marked of a difference).

**Comment by qwertychouskie** [comment]
> BTW, I figured out what caused the significant regression in app deployment times in the Into The Deep season, it's an Android Studio regression in 2024.1.2. Partial deploys worked in 2024.1.1.13, but do not in 2024.1.2.13. I have a bug report open here: https://issuetracker.google.com/issues/442347342
> 
> Hopefully this gets fixed upstream in a timely manner, but any APK size reduction we can do helps mitigate the issue to a certain degree (and even when the upstream partial deploy issue is fixed, a smaller APK is still beneficial, though less marked of a difference).

Good news, this has been fixed upstream in the new Canary build!  (https://developer.android.com/studio/preview)  Now to see if they will backport the fix to the stable series...

**Comment by polarbub** [comment]
Another more invasive optimization that can be done is to completely disable OnBot and Blocks. It cuts down the size a lot.

I did this by using [ExtractedRC](https://github.com/OpenFTC/Extracted-RC). I removed the dependency on the OnBotJava and Blocks projects from gradle, moved `assets/java` from RobotCore to OnBotJava, and removed the OnBotJava service from RobotCore's `AndroidManifest.xml`. I also had to patch `FtcRobotControllerActivity.java` to not reference OnBot or Blocks.

**Comment by Windwoes** [comment]
> Another more invasive optimization that can be done is to completely disable OnBot and Blocks. It cuts down the size a lot.
> 
> I did this by using [ExtractedRC](https://github.com/OpenFTC/Extracted-RC). I removed the dependency on the OnBotJava and Blocks projects from gradle, moved `assets/java` from RobotCore to OnBotJava, and removed the OnBotJava service from RobotCore's `AndroidManifest.xml`. I also had to patch `FtcRobotControllerActivity.java` to not reference OnBot or Blocks.

You've basically just described [OpenRC Turbo](https://github.com/OpenFTC/OpenRC-Turbo). After interest dropped off I stopped maintaining it, though.

**Comment by polarbub** [comment]
Yeah, OpenRC Turbo is where I got the idea from, though I haven't been able to get down to minuscule 7.6 MB that OpenRC was able to get to. I'm at 20.5 MB, but all my .so libraries are packaged in the APK. Still at that size it cuts the upload time more than in half.
